### PR TITLE
Feature/trucks

### PR DESCRIPTION
**Pull request by justaname94**
Add models, storage and functions for trucks. The functions added are for creating, getting and updating a single truck

### CODE REVIEW

**Comment by pablocampogo** [comment]
What are all the PNGs for? You should explain somewhere in the code their usage

File: shared/truck-models/truck.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	CompletedTrips    int          `json:"completed_trips"`
+	Available         bool         `json:"available"`
+	Type              TruckType    `json:"type"`
+	Location          trips.Region `json:"location"`
+	CreationDate      time.Time    `json:"creation_date"`
+	UpdateDate        time.Time    `json:"update_date"`
+}
+
+// NewTruck returns a cleaned Truck structure with given values
+func NewTruck(truck Truck) (*Truck, error) {
```
Comment (pablocampogo) [inline shared/truck-models/truck.go:86]: There is a smarter way of doing this  jeje, but I will do it myself

File: trucks/functions/create-truck/main.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+func apiGatewayHandler(ctx context.Context, request events.APIGatewayProxyRequest) (*events.APIGatewayProxyResponse, error) {
+	body, err := url.ParseQuery(request.Body)
+	if err != nil {
+		return apigateway.LogAndReturnError(err), nil
+	}
+
+	year, err := strconv.Atoi(body.Get("year"))
+	if err != nil {
+		return apigateway.LogAndReturnError(models.ErrInvalidYear), nil
```
Comment (pablocampogo) [inline trucks/functions/create-truck/main.go:26]: This is a 400, client did not send us a number

File: trucks/functions/create-truck/main.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	}
+
+	mileague, err := strconv.Atoi(body.Get("mileague"))
+	if err != nil {
+		return apigateway.LogAndReturnError(models.ErrInvalidMileague), nil
```
Comment (pablocampogo) [inline trucks/functions/create-truck/main.go:31]: This is also a 400

File: trucks/functions/get-truck/main.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+func apiGatewayHandler(ctx context.Context, request events.APIGatewayProxyRequest) (*events.APIGatewayProxyResponse, error) {
+	body, err := url.ParseQuery(request.Body)
+	if err != nil {
+		return apigateway.LogAndReturnError(err), nil
+	}
+
+	truck, err := storage.LoadTruck(ctx, body.Get("truck_id"))
+	if errors.Is(err, storage.ErrTruckNotFound) {
+		return apigateway.NewErrorResponse(404, err), nil
+	}
```
Comment (pablocampogo) [inline trucks/functions/get-truck/main.go:23]: Add a blank line after this

File: trucks/functions/update-truck/main.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+func apiGatewayHandler(ctx context.Context, request events.APIGatewayProxyRequest) (*events.APIGatewayProxyResponse, error) {
+	body, err := url.ParseQuery(request.Body)
+	if err != nil {
+		return apigateway.LogAndReturnError(err), nil
+	}
+
+	truck, err := storage.LoadTruck(ctx, body.Get("truck_id"))
+	if errors.Is(err, storage.ErrTruckNotFound) {
+		return apigateway.NewErrorResponse(404, err), nil
+	}
```
Comment (pablocampogo) [inline trucks/functions/update-truck/main.go:27]: Blank line after this

File: trucks/functions/update-truck/main.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+	if body.Get("model") != "" {
+		truck.Model = body.Get("model")
+	}
+
+	if body.Get("year") != "" {
+		year, err := strconv.Atoi(body.Get("year"))
+		if err != nil {
+			return apigateway.LogAndReturnError(models.ErrInvalidYear), nil
```
Comment (pablocampogo) [inline trucks/functions/update-truck/main.go:47]: This is a 400

File: trucks/functions/update-truck/main.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+		}
+
+		truck.Year = year
+	}
+
+	if body.Get("mileague") != "" {
+		mileague, err := strconv.Atoi(body.Get("mileague"))
+		if err != nil {
+			return apigateway.LogAndReturnError(models.ErrInvalidMileague), nil
```
Comment (pablocampogo) [inline trucks/functions/update-truck/main.go:56]: This is a 400

File: trucks/functions/update-truck/main.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+		}
+
+		truck.Mileague = mileague
+	}
+
+	if body.Get("available") != "" {
+		available, err := strconv.ParseBool(body.Get("available"))
+		if err != nil {
+			return apigateway.LogAndReturnError(models.ErrInvalidAvailable), nil
```
Comment (pablocampogo) [inline trucks/functions/update-truck/main.go:65]: This is a 400

File: trucks/functions/update-truck/main.go [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+		}
+
+		truck.Available = available
+	}
+
+	if body.Get("type") != "" {
+		truckType := models.TruckType(body.Get("type"))
+		_, ok := models.ValidTruckTypes[truckType]
+		if !ok {
+			return apigateway.LogAndReturnError(models.ErrInvalidTruckType), nil
```
Comment (pablocampogo) [inline trucks/functions/update-truck/main.go:75]: This is a 400

File: trucks/functions/update-truck/main.go [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+		}
+		truck.Type = truckType
+	}
+
+	if body.Get("location") != "" {
+		location := trips.Region(body.Get("location"))
+		_, ok := trips.ValidRegions[location]
+		if !ok {
+			return apigateway.LogAndReturnError(models.ErrInvalidLocation), nil
```
Comment (pablocampogo) [inline trucks/functions/update-truck/main.go:84]: This is a 400

File: trucks/functions/create-truck/main.go [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+		Model:             body.Get("model"),
+		Year:              year,
+		Mileague:          mileague,
+		Type:              models.TruckType(body.Get("type")),
+		Location:          trips.Region(body.Get("location")),
+	}
+
+	truck, err := models.NewTruck(*truckParam)
+	if err != nil {
+		return apigateway.LogAndReturnError(err), nil
```
Comment (pablocampogo) [inline trucks/functions/create-truck/main.go:55]: This is a 400

**Review by pablocampogo — CHANGES_REQUESTED** [review]
Solve all comments but the one in NewTruck and delete PNGs unless they are necessary

File: shared/truck-models/truck.go (hunk 1 above)
Comment (justaname94) [reply shared/truck-models/truck.go:86]: Lol, I asked you if there was a better way of doing it. But awesome, I'll see how you do it

**Review by pablocampogo — APPROVED** [review]
Great job!

**Pull request merged**